tests/drivers/input/kbd_matrix/src/main.c

```diff
@@ -0,0 +1,399 @@
+{
+	return state.rows[state.col];
+}
+
+static void test_set_detect_mode(const struct device *dev, bool enabled)
+{
+	TC_PRINT("detect mode: enabled=%d\n", enabled);
+	state.detect_mode = enabled;
+}
+
+static const struct input_kbd_matrix_api test_api = {
+	.drive_column = test_drive_column,
+	.read_row = test_read_row,
+	.set_detect_mode = test_set_detect_mode,
+};
+
+INPUT_KBD_MATRIX_DT_DEFINE(TEST_KBD_SCAN_NODE);
+
+static const struct input_kbd_matrix_common_config
+	test_cfg = INPUT_KBD_MATRIX_DT_COMMON_CONFIG_INIT(
+			TEST_KBD_SCAN_NODE, &test_api);
+
+static struct input_kbd_matrix_common_data test_data;
+
+DEVICE_DT_DEFINE(TEST_KBD_SCAN_NODE, input_kbd_matrix_common_init, NULL,
+		 &test_data, &test_cfg,
+		 POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE, NULL);
+
+static const struct device *const test_dev = DEVICE_DT_GET(TEST_KBD_SCAN_NODE);
+
+/* support stuff */
+
+static const struct device *column_hook_last_dev;
+static int column_hook_last_col;
+
+void input_kbd_matrix_drive_column_hook(const struct device *dev, int col)
+{
+	column_hook_last_dev = dev;
+	column_hook_last_col = col;
+}
+
+static void state_set_rows(kbd_row_t r0, kbd_row_t r1, kbd_row_t r2)
+{
+	memcpy(&state.rows, (kbd_row_t[]){r0, r1, r2}, sizeof(state.rows));
+	TC_PRINT("set state [" PRIkbdrow " " PRIkbdrow " " PRIkbdrow "]\n", r0, r1, r2);
+}
+
+static struct {
+	int row;
+	int col;
+	int val;
+	int event_count;
+} test_event_data;
+
+static int last_checked_event_count;
+
+#define assert_no_new_events() \
+	zassert_equal(last_checked_event_count, test_event_data.event_count);
+
+#define  assert_new_event(_row, _col, _val) { \
+	last_checked_event_count++; \
+	zassert_equal(last_checked_event_count, test_event_data.event_count); \
+	zassert_equal(_row, test_event_data.row); \
+	zassert_equal(_col, test_event_data.col); \
+	zassert_equal(_val, test_event_data.val); \
+}
+
+static void test_cb(struct input_event *evt)
+{
+	static int row, col, val;
+
+	switch (evt->code) {
+	case INPUT_ABS_X:
+		col = evt->value;
+		break;
+	case INPUT_ABS_Y:
+		row = evt->value;
+		break;
+	case INPUT_BTN_TOUCH:
+		val = evt->value;
+		break;
+	}
+
+	if (evt->sync) {
+		test_event_data.row = row;
+		test_event_data.col = col;
+		test_event_data.val = val;
+		test_event_data.event_count++;
+		TC_PRINT("input event: count=%d row=%d col=%d val=%d\n",
+			 test_event_data.event_count, row, col, val);
+	}
+}
+INPUT_CALLBACK_DEFINE(test_dev, test_cb);
+
+#define WAIT_FOR_IDLE_TIMEOUT_US (5 * USEC_PER_SEC)
+
+static void kbd_scan_wait_for_idle(void)
+{
+	bool to;
+
+	to = WAIT_FOR(state.detect_mode,
+		      WAIT_FOR_IDLE_TIMEOUT_US,
+		      k_sleep(K_MSEC(100)));
+
+	zassert_true(to, "timeout waiting for idle state");
+}
+
+/* actual tests */
+
+/* no event before debounce time, event after */
+ZTEST(kbd_scan, test_kbd_scan)
+{
+	const struct input_kbd_matrix_common_config *cfg = test_dev->config;
+
+	input_kbd_matrix_poll_start(test_dev);
+
+	state_set_rows(0x00, 0x04, 0x00);
```

Comment:
Using the bit mask makes it clearer that you're asserting row 2 on column 1.
```suggestion
	state_set_rows(0x00, BIT(2), 0x00);
```

Reply:
Cool, changed them all.